### PR TITLE
[PAY-1139] Throttle calls to fetchMessages on web scroll

### PR DESCRIPTION
**Pull request by piazzatron**
### Description

- On web, throttle calls to fetchMessages once per 500ms.
- Fixes a bug the existing web scroll handlers where we failed to check if `prev_count > 0` when fetching at the top of the list, leading to spamming comms:
```
// old version:
if (
            chatId &&
            isScrolledToTop(e.currentTarget) &&
            chat?.messagesStatus !== Status.LOADING
          ) {
            // Fetch more messages when user reaches the top
            dispatch(fetchMoreMessages({ chatId }))
          }
```
Mobile was performing this check correctly, so I took this as opportunity to consolidate this code.

On Throttle vs Debounce:
- I picked throttle because debounce will only call the final invocation of a function; we want to call this periodically as a user scrolls if they're scrolling through a long list of messages.

### Dragons

*Is there anything the reviewer should be on the lookout for? Are there any dangerous changes?*

### How Has This Been Tested?

Locally

*Please describe the tests that you ran to verify your changes. Provide repro instructions & any configuration.*

### How will this change be monitored?

*For features that are critical or could fail silently please describe the monitoring/alerting being added.*

### Feature Flags ###

*Are all new features properly feature flagged? Describe added feature flags.*

### CODE REVIEW

File: packages/common/src/utils/chatUtils.ts [hunk 1]
```diff
@@ -45,3 +47,14 @@ export const isEarliestUnread = ({
   const isAuthor = message.sender_user_id === currentUserId
   return isUnread && !isPreviousMessageUnread && !isAuthor
 }
+
+/**
+ * Can only fetch more messages if it's not in a loading state and
+ * there are previous messages to fetch
+ */
+export const chatCanFetchMoreMessages = (
```
Comment (rickyrombo) [inline packages/common/src/utils/chatUtils.ts:55]: very nit but this could even take a chatID and just be a selector?

File: packages/web/src/pages/chat-page/components/ChatMessageList.tsx [hunk 2]
```diff
@@ -46,13 +49,14 @@ type ChatMessageListProps = ComponentPropsWithoutRef<'div'> & {
 
 const SCROLL_TOP_THRESHOLD = 800
 const SCROLL_BOTTOM_THRESHOLD = 32
+const THROTTLE_DURATION_MS = 500
 
-const isScrolledToBottom = (element: HTMLElement) => {
+const isScrolledNearBottom = (element: HTMLElement) => {
```
Comment (rickyrombo) [inline packages/web/src/pages/chat-page/components/ChatMessageList.tsx:54]: ty

**Comment by audius-infra** [comment]
Preview this change https://demo.audius.co/piazz-debounce-fix

**Review by rickyrombo — APPROVED** [review]

File: packages/web/src/pages/chat-page/components/ChatMessageList.tsx [hunk 3]
```diff
@@ -83,27 +87,51 @@ export const ChatMessageList = forwardRef<HTMLDivElement, ChatMessageListProps>(
...
-      },
-      [dispatch, chatId, chat?.messagesStatus]
+    // Memoize the creation of throttled scroll handler, to avoid
+    // creating a new throttled function each time and because useCallback
+    // doesn't like receiving a non-inlined fn
+    // https://dmitripavlutin.com/react-throttle-debounce/
+    const handleScroll = useMemo(
+      () =>
+        throttle(
+          (e: UIEvent<HTMLDivElement>) => {
```
[the remark this answers is not in the document]
Comment (piazzatron) [reply packages/web/src/pages/chat-page/components/ChatMessageList.tsx:97]: lolll this is exactly how I had it initially and then changed it because I was running into some very strange issues... let me try again

File: packages/common/src/utils/chatUtils.ts (hunk 1 above)
Comment (piazzatron) [reply packages/common/src/utils/chatUtils.ts:55]: discussed, leaving as is
